pkg/keystorage/keystorage.go

```diff
@@ -0,0 +1,193 @@
+package keystorage
+
+import (
+	"crypto/hmac"
+	"crypto/sha256"
+	"crypto/subtle"
+	"fmt"
+	"sync"
+
+	"github.com/ProtonMail/gopenpgp/v2/helper"
+
+	"github.com/cosi-project/runtime/api/key_storage"
+)
+
+// KeyStorage is a key storage that can be used to store and retrieve the master key.
+//
+//nolint:govet
+type KeyStorage struct {
+	mx         sync.Mutex
+	underlying key_storage.Storage
+}
+
+// SetMasterKey sets the master key for the key storage, encrypts it using public key and stores it in slot id.
+func (ks *KeyStorage) SetMasterKey(masterKey []byte, slotID, slotPublicKey string) error {
+	switch {
+	case len(masterKey) != 32:
+		return fmt.Errorf("master key can only be 32 bytes long")
+	case slotID == "":
+		return fmt.Errorf("slot id cannot be empty")
+	case slotPublicKey == "":
+		return fmt.Errorf("slot public key cannot be empty")
+	}
+
+	ks.mx.Lock()
+	defer ks.mx.Unlock()
+
+	if !isZero(&ks.underlying) {
+		return fmt.Errorf("key storage is already initialized")
+	}
+
+	encryptedSlot, err := helper.EncryptBinaryMessageArmored(slotPublicKey, masterKey)
+	if err != nil {
+		return fmt.Errorf("failed to encrypt slot '%s': %w", slotID, err)
+	}
+
+	ks.underlying.StorageVersion = key_storage.StorageVersion_STORAGE_VERSION_1
+	ks.underlying.KeySlot = map[string]*key_storage.KeySlot{
+		slotID: {
+			Algorithm:    key_storage.Algorithm_PGP_AES_GCM_256,
+			EncryptedKey: []byte(encryptedSlot),
+		},
+	}
+	ks.underlying.KeysHmacHash = ks.hashSlots(masterKey)
+
+	return nil
+}
+
+// SetKey sets the new key slot with the given slot id and public key using previus slot and its private key.
+// It cannot be used to update an existing key slot.
+// It is required to call SetMasterKey() or UnmarshalBinary() to initialize the key storage first.
+func (ks *KeyStorage) SetKey(newSlotID, newSlotPublicKey, oldSlotID, oldSlotPrivateKey string) error {
+	switch {
+	case newSlotID == "":
+		return fmt.Errorf("new slot id cannot be empty")
+	case newSlotPublicKey == "":
+		return fmt.Errorf("new slot public key cannot be empty")
+	}
+
+	ks.mx.Lock()
+	defer ks.mx.Unlock()
+
+	if newSlot := ks.underlying.GetKeySlot()[newSlotID]; newSlot != nil {
+		return fmt.Errorf("new slot '%s' already exists", newSlotID)
+	}
+
+	masterKey, err := ks.getKey(oldSlotID, oldSlotPrivateKey)
+	if err != nil {
+		return err
+	}
+
+	encryptedSlot, err := helper.EncryptBinaryMessageArmored(newSlotPublicKey, masterKey)
+	if err != nil {
+		return fmt.Errorf("failed to encrypt slot '%s': %w", newSlotID, err)
+	}
+
+	ks.underlying.KeySlot[newSlotID] = &key_storage.KeySlot{
+		Algorithm:    key_storage.Algorithm_PGP_AES_GCM_256,
+		EncryptedKey: []byte(encryptedSlot),
+	}
+
+	ks.underlying.KeysHmacHash = ks.hashSlots(masterKey)
+
+	return nil
+}
+
+// GetKey returns the attempts to decrypt the key slot with the given private key and return the master key.
+func (ks *KeyStorage) GetKey(slotID, slotPrivateKey string) ([]byte, error) {
```

Comment:
I think this returns the master key, to distinguish from `SetKey`, which operates on key slots

```suggestion
func (ks *KeyStorage) GetMasterKey(slotID, slotPrivateKey string) ([]byte, error) {
```

Reply:
Done!